### PR TITLE
[KOGITO-1919] - Allow user to specify the artifact and group id when …

### PR DESCRIPTION
**Pull request by vaibhavjainwiz**
…doing builds from archetype

Jira Issue : https://issues.redhat.com/browse/KOGITO-1919

Many thanks for submiting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](CONTRIBUTING.MD#sending-a-pull-request)
- [x] Pull Request title is properly formatted: `[KOGITO-XYZ] Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Your feature/bug fix has a unit test that verifies it
- [x] You've tested the new feature/bug fix in an actual OpenShift cluster

### CODE REVIEW

**Comment by spolti** [comment]
@vaibhavjainwiz could you please check the errors? Both jobs are failing.

File: cmd/kogito/command/deploy/deploy_service_test.go [hunk 1]
```diff
@@ -331,3 +331,71 @@ func Test_WithManagementConsole(t *testing.T) {
 	assert.Contains(t, lines, "successfully created")
 	assert.Contains(t, lines, mgmtConsoleURI)
 }
+
+func Test_DeployCmd_WithCustomArtifactDetails(t *testing.T) {
+	ns := t.Name()
+	groupId := "com.mycompany"
+	artifactId := "testproject"
+	version := "2.0-SNAPSHOT"
+	fileLocation, _ := url.Parse("https:ss//raw.githubusercontent.com/kiegroup/kogito-examples/master/dmn-quarkus-example/src/main/resources/Traffic Violation.dmnsd")
```
Comment (sutaakar) [inline cmd/kogito/command/deploy/deploy_service_test.go:340]:
```suggestion
	fileLocation, _ := url.Parse("https://raw.githubusercontent.com/kiegroup/kogito-examples/master/dmn-quarkus-example/src/main/resources/Traffic Violation.dmn")
```

File: cmd/kogito/command/deploy/deploy_service_test.go [hunk 2]
```diff
@@ -331,3 +331,71 @@ func Test_WithManagementConsole(t *testing.T) {
...
+	assert.Equal(t, artifactId, kogitoApp.Spec.Build.Artifact.ArtifactId)
+	assert.Equal(t, version, kogitoApp.Spec.Build.Artifact.Version)
+}
+
+func Test_DeployCmd_WithDefaultArtifactDetails(t *testing.T) {
+	ns := t.Name()
+	groupId := "com.company"
+	artifactId := "project"
+	version := "1.0-SNAPSHOT"
+	fileLocation, _ := url.Parse("https:ss//raw.githubusercontent.com/kiegroup/kogito-examples/master/dmn-quarkus-example/src/main/resources/Traffic Violation.dmnsd")
```
Comment (sutaakar) [inline cmd/kogito/command/deploy/deploy_service_test.go:374]:
```suggestion
	fileLocation, _ := url.Parse("https://raw.githubusercontent.com/kiegroup/kogito-examples/master/dmn-quarkus-example/src/main/resources/Traffic Violation.dmn")
```

File: cmd/kogito/command/deploy/deploy_service_test.go [hunk 3]
```diff
@@ -331,3 +331,71 @@ func Test_WithManagementConsole(t *testing.T) {
 	assert.Contains(t, lines, "successfully created")
 	assert.Contains(t, lines, mgmtConsoleURI)
 }
+
+func Test_DeployCmd_WithCustomArtifactDetails(t *testing.T) {
+	ns := t.Name()
+	groupId := "com.mycompany"
+	artifactId := "testproject"
+	version := "2.0-SNAPSHOT"
+	fileLocation, _ := url.Parse("https://raw.githubusercontent.com/kiegroup/kogito-examples/master/dmn-quarkus-example/src/main/resources/Traffic Violation.dmn")
```
Comment (ricardozanini) [inline cmd/kogito/command/deploy/deploy_service_test.go:340]: Instead, please bring the file to the `testdata` directory and use it from there.

File: cmd/kogito/command/deploy/deploy_service_test.go [hunk 4]
```diff
@@ -331,3 +331,71 @@ func Test_WithManagementConsole(t *testing.T) {
...
+	assert.Equal(t, artifactId, kogitoApp.Spec.Build.Artifact.ArtifactId)
+	assert.Equal(t, version, kogitoApp.Spec.Build.Artifact.Version)
+}
+
+func Test_DeployCmd_WithDefaultArtifactDetails(t *testing.T) {
+	ns := t.Name()
+	groupId := "com.company"
+	artifactId := "project"
+	version := "1.0-SNAPSHOT"
+	fileLocation, _ := url.Parse("https:ss//raw.githubusercontent.com/kiegroup/kogito-examples/master/dmn-quarkus-example/src/main/resources/Traffic Violation.dmnsd")
```
Comment (ricardozanini) [inline cmd/kogito/command/deploy/deploy_service_test.go:374]: Same here regarding the `testdata` directory.

File: pkg/apis/app/v1alpha1/kogitoapp_types.go [hunk 5]
```diff
@@ -136,6 +136,9 @@ type KogitoAppBuildObject struct {
 	Resources corev1.ResourceRequirements `json:"resources,omitempty"`
 	// Internal Maven Mirror to be used during source-to-image builds to considerably increase build speed
 	MavenMirrorURL string `json:"mavenMirrorURL,omitempty"`
+	//Artifact contains final generated project artifact
+	// + optional
+	Artifact Artifact `json:"artifact,omitempty"`
```
Comment (ricardozanini) [inline pkg/apis/app/v1alpha1/kogitoapp_types.go:141]: Please add the `operator-sdk` flags to the attributes of this type.

File: pkg/apis/app/v1alpha1/kogitoservices_types.go [hunk 6]
```diff
@@ -225,7 +225,9 @@ func (k *KogitoServiceSpec) AddResourceLimit(name, value string) {
 func (k *KogitoServiceSpec) GetDeploymentLabels() map[string]string { return k.DeploymentLabels }
 
 // SetDeploymentLabels ...
-func (k *KogitoServiceSpec) SetDeploymentLabels(labels map[string]string) { k.DeploymentLabels = labels }
+func (k *KogitoServiceSpec) SetDeploymentLabels(labels map[string]string) {
```
Comment (ricardozanini) [inline pkg/apis/app/v1alpha1/kogitoservices_types.go:228]: Please do not change this inline function.

File: pkg/controller/kogitoapp/resource/build_config_s2i.go [hunk 7]
```diff
@@ -117,6 +120,21 @@ func setBCS2IStrategy(kogitoApp *v1alpha1.KogitoApp, buildConfig *buildv1.BuildC
 		envs = framework.EnvOverride(envs, corev1.EnvVar{Name: mavenMirrorURLEnvVar, Value: kogitoApp.Spec.Build.MavenMirrorURL})
 	}
 
+	if len(kogitoApp.Spec.Build.Artifact.GroupId) > 0 {
+		log.Infof("Setting final generated artifact group id %s", kogitoApp.Spec.Build.Artifact.GroupId)
```
Comment (ricardozanini) [inline pkg/controller/kogitoapp/resource/build_config_s2i.go:124]: These logs can be `debug` instead.

File: test/framework/openshift.go [hunk 8]
```diff
@@ -92,7 +92,7 @@ func WaitForDeploymentConfigRunning(namespace, dcName string, podNb int, timeout
 				}
 
 				GetLogger(namespace).Debugf("Deployment config has %d pods\n", len(pods.Items))
-				return len(pods.Items) == podNb && CheckPodsAreRunning(pods), nil
+				return len(pods.Items) == podNb && CheckPodsAreReady(pods), nil
```
Comment (ricardozanini) [inline test/framework/openshift.go:95]: Why this change in this PR?

**Comment by radtriste** [comment]
depends on https://github.com/kiegroup/kogito-images/pull/164

File: cmd/kogito/command/deploy/deploy_service.go [hunk 9]
```diff
@@ -163,6 +166,9 @@ func (i *deployCommand) InitHook() {
 	i.command.Flags().StringVar(&i.flags.imageVersion, "image-version", "", "Image version for standard Kogito build images. Ignored if a custom image is set for image-s2i or image-runtime.")
 	i.command.Flags().StringVar(&i.flags.mavenMirrorURL, "maven-mirror-url", "", "Internal Maven Mirror to be used during source-to-image builds to considerably increase build speed, e.g: https://my.internal.nexus/content/group/public")
 	i.command.Flags().BoolVar(&i.flags.enableIstio, "enable-istio", false, "Enable Istio integration by annotating the Kogito service pods with the right value for Istio controller to inject sidecars on it. Defaults to false")
+	i.command.Flags().StringVar(&i.flags.projectGroupId, "project-group-id", "com.company", "Indicates the unique identifier of the organization or group that created the project.")
```
Comment (radtriste) [inline cmd/kogito/command/deploy/deploy_service.go:169]: Shouldn't we have empty values by default like it was before ?

File: pkg/apis/app/v1alpha1/kogitoapp_types.go [hunk 10]
```diff
@@ -136,6 +136,9 @@ type KogitoAppBuildObject struct {
 	Resources corev1.ResourceRequirements `json:"resources,omitempty"`
 	// Internal Maven Mirror to be used during source-to-image builds to considerably increase build speed
 	MavenMirrorURL string `json:"mavenMirrorURL,omitempty"`
+	//Artifact contains final generated project artifact
```
Comment (radtriste) [inline pkg/apis/app/v1alpha1/kogitoapp_types.go:139]:
```suggestion
	// Artifact contains override information for building the Maven artifact
```

File: pkg/controller/kogitoapp/resource/build_config_s2i.go [hunk 11]
```diff
@@ -36,6 +36,9 @@ const (
 	buildS2IlimitCPUEnvVarKey    = "LIMIT_CPU"
 	buildS2IlimitMemoryEnvVarKey = "LIMIT_MEMORY"
 	mavenMirrorURLEnvVar         = "MAVEN_MIRROR_URL"
+	projectGroupId               = "PROJECT_GROUP_ID"
+	projectArtifactId            = "PROJECT_ARTIFACT_ID"
+	projectVersion               = "PROJECT_VERSION"
```
Comment (radtriste) [inline pkg/controller/kogitoapp/resource/build_config_s2i.go:41]: Please end the var names with `EnvVar` suffix

File: test/framework/openshift.go (hunk 8 above)
Comment (radtriste) [reply test/framework/openshift.go:95]: this has been already corrected normally on master. A simple rebase should have taken it.

File: pkg/controller/kogitoapp/resource/build_config_s2i.go (hunk 11 above)
Comment (spolti) [reply pkg/controller/kogitoapp/resource/build_config_s2i.go:41]: I would suggest to use mavenGroupIdEnvVar and so on.

File: pkg/controller/kogitoapp/resource/build_config_s2i.go [hunk 12]
```diff
@@ -117,6 +120,21 @@ func setBCS2IStrategy(kogitoApp *v1alpha1.KogitoApp, buildConfig *buildv1.BuildC
 		envs = framework.EnvOverride(envs, corev1.EnvVar{Name: mavenMirrorURLEnvVar, Value: kogitoApp.Spec.Build.MavenMirrorURL})
 	}
 
+	if len(kogitoApp.Spec.Build.Artifact.GroupId) > 0 {
```
Comment (spolti) [inline pkg/controller/kogitoapp/resource/build_config_s2i.go:123]: I would add a extra if here checking if the build is really from file and a build using the archetype needed.

Comment (sutaakar) [reply pkg/controller/kogitoapp/resource/build_config_s2i.go:123]: Actually there are two options when artifact configuration can be user:
- build from file
- build from URL pointing to file or dir with source file

Comment (vaibhavjainwiz) [reply pkg/controller/kogitoapp/resource/build_config_s2i.go:123]: If URL is supplied we have no way of identifying where does it lead to. This type of checks are already implemented in builder images where we checkout complete git repo from provided url and verify whether build using the archetype needed. Currently not implementing any checks here

File: cmd/kogito/command/deploy/deploy_service.go (hunk 9 above)
Comment (vaibhavjainwiz) [reply cmd/kogito/command/deploy/deploy_service.go]: set default value as empty

File: pkg/apis/app/v1alpha1/kogitoapp_types.go (hunk 5 above)
Comment (vaibhavjainwiz) [reply pkg/apis/app/v1alpha1/kogitoapp_types.go:143]: done

File: pkg/apis/app/v1alpha1/kogitoservices_types.go (hunk 6 above)
Comment (vaibhavjainwiz) [reply pkg/apis/app/v1alpha1/kogitoservices_types.go:228]: It changed due to code formating when I run `make build`, reverted it now

File: cmd/kogito/command/deploy/deploy_service_test.go (hunk 3 above)
Comment (vaibhavjainwiz) [reply cmd/kogito/command/deploy/deploy_service_test.go]: actual fileLocation never refered while running BATS, still I replace to testdata location

File: pkg/apis/app/v1alpha1/kogitoapp_types.go (hunk 5 above)
Comment (vaibhavjainwiz) [reply pkg/apis/app/v1alpha1/kogitoapp_types.go:143]: done

File: pkg/controller/kogitoapp/resource/build_config_s2i.go (hunk 7 above)
Comment (vaibhavjainwiz) [reply pkg/controller/kogitoapp/resource/build_config_s2i.go]: change to debug level

File: pkg/apis/app/v1alpha1/kogitoapp_types.go (hunk 10 above)
Comment (vaibhavjainwiz) [reply pkg/apis/app/v1alpha1/kogitoapp_types.go]: description updated

File: pkg/controller/kogitoapp/resource/build_config_s2i.go (hunk 11 above)
Comment (vaibhavjainwiz) [reply pkg/controller/kogitoapp/resource/build_config_s2i.go]: changed name to `mavenGroupIdEnvVar`

File: pkg/apis/app/v1alpha1/kogitoservices_types.go (hunk 6 above)
Comment (vaibhavjainwiz) [reply pkg/apis/app/v1alpha1/kogitoservices_types.go:228]: done

File: cmd/kogito/command/deploy/deploy_service_test.go (hunk 4 above)
Comment (vaibhavjainwiz) [reply cmd/kogito/command/deploy/deploy_service_test.go]: edit test to remove remote file location

File: pkg/controller/kogitoapp/resource/build_config_s2i.go (hunk 12 above)
Comment (spolti) [reply pkg/controller/kogitoapp/resource/build_config_s2i.go:123]: don't you agree that you are setting unneeded things parameters?

File: pkg/apis/app/v1alpha1/kogitoapp_types.go [hunk 13]
```diff
@@ -324,3 +329,19 @@ type KogitoAppList struct {
 func init() {
 	SchemeBuilder.Register(&KogitoApp{}, &KogitoAppList{})
 }
+
+//Artifact contains final generated project artifact
+type Artifact struct {
```
Comment (ricardozanini) [inline pkg/apis/app/v1alpha1/kogitoapp_types.go:334]: The attributes of this structure need to be annotated as well. Sorry if my last comment wasn't clear.

Comment (vaibhavjainwiz) [reply pkg/apis/app/v1alpha1/kogitoapp_types.go:337]: done

Comment (ricardozanini) [reply pkg/apis/app/v1alpha1/kogitoapp_types.go:337]: @vaibhavjainwiz the attributes of the structure is `GroupId`, `ArtifactId` and `Version`.
